pkg/ebpf/map.go

```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
 	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
 	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
 
+	path := bpf.MapPath(m.spec.Name)
+
 	newMap, err := ciliumebpf.NewMapWithOptions(m.spec, opts)
 	if err != nil {
-		return fmt.Errorf("unable to create map: %w", err)
+		if !errors.Is(err, ciliumebpf.ErrMapIncompatible) {
+			return fmt.Errorf("unable to create map: %w", err)
+		}
+
+		// There already exists a pinned map but it has a different
+		// configuration (e.g different type, k/v size or flags).
+		// Try to delete and recreate it.
+
+		log.WithField("map", m.spec.Name).
+			Warn("cannot load pinned map as it's incompatible with the spec provided. Deleting and recreating it.")
```

Comment:
A few comments here:

* The current implementation, via `objCheck()`, logs specifically which attributes are incompatible. This means that we have some chance of understanding what the incompatibility is and this can go a long way to fixing unintended breakages introduced by developers. Without that, since this code is deleting the old map, it can be quite difficult to understand exactly what is going on because the old map is gone by the time you go to inspect it.
* The current log uses the wording `Removing map to allow for property upgrade (expect map data loss)`, basically to try to raise the alarm that this is really not expected. We can maybe argue if it's too alarming given that I believe only CT map deletion would cause actual packet impact, but I wanted to at least check whether there was any discussion about this? Even if we used the exact same wording, we should be able to differentiate log messages from the old package vs. the new package by checking the `subsys` field (should be `ebpf` for the new version, or `bpf` for the old one).
* nit, while errors should start with lower case (since they are often concatenated with other messages), log messages usually start with a capital case.

Reply:
I updated the message to include the error from `cilium/ebpf` that indicates which attribute is incompatible with the spec and reintroduced the wording of the previous log. The user should now get something like:

```
level=warning msg="Removing map to allow for property upgrade (expect map data loss): expected flags 1, got 0: map's spec is incompatible with pinned map" map=cilium_metrics subsys=bpf
```